File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/tasks/HabitViewHolder.java
package com.habitrpg.android.habitica.ui.viewHolders.tasks;

import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.events.HabitScoreEvent;
import com.habitrpg.android.habitica.models.tasks.Task;

import org.greenrobot.eventbus.EventBus;

import butterknife.BindView;
import butterknife.OnClick;

public class HabitViewHolder extends BaseTaskViewHolder {

    @BindView(R.id.btnPlusWrapper)
    FrameLayout btnPlusWrapper;
    @BindView(R.id.btnPlusIconView)
    ImageView btnPlusIconView;
    @BindView(R.id.btnPlus)
    Button btnPlus;

    @BindView(R.id.btnMinusWrapper)
    FrameLayout btnMinusWrapper;
    @BindView(R.id.btnMinusIconView)
    ImageView btnMinusIconView;
    @BindView(R.id.btnMinus)
    Button btnMinus;

    @BindView(R.id.streakTextView)
    TextView streakTextView;

    public HabitViewHolder(View itemView) {
        super(itemView);
    }

    @Override
    public void bindHolder(Task newTask, int position) {
        this.task = newTask;
        if (newTask.up) {
            this.btnPlusWrapper.setBackgroundResource(newTask.getLightTaskColor());
            if (newTask.getLightTaskColor() == R.color.yellow_100) {
                this.btnPlusIconView.setImageResource(R.drawable.habit_plus_yellow);
            } else {
                this.btnPlusIconView.setImageResource(R.drawable.habit_plus);
            }
            this.btnPlus.setVisibility(View.VISIBLE);
            this.btnPlus.setClickable(true);
        } else {
            this.btnPlusWrapper.setBackgroundResource(R.color.habit_inactive_gray);
            this.btnPlusIconView.setImageResource(R.drawable.habit_plus_disabled);
            this.btnPlus.setVisibility(View.GONE);
            this.btnPlus.setClickable(false);
        }

        if (newTask.down) {
            this.btnMinusWrapper.setBackgroundResource(newTask.getLightTaskColor());
            if (newTask.getLightTaskColor() == R.color.yellow_100) {
                this.btnMinusIconView.setImageResource(R.drawable.habit_minus_yellow);
            } else {
                this.btnMinusIconView.setImageResource(R.drawable.habit_minus);
            }
            this.btnMinus.setVisibility(View.VISIBLE);
            this.btnMinus.setClickable(true);
        } else {
            this.btnMinusWrapper.setBackgroundResource(R.color.habit_inactive_gray);
            this.btnMinusIconView.setImageResource(R.drawable.habit_minus_disabled);
            this.btnMinus.setVisibility(View.GONE);
            this.btnMinus.setClickable(false);
        }

        String streakString = "";
        if (newTask.counterUp > 0 && newTask.counterDown > 0) {
            streakString = streakString + "+" + String.valueOf(task.counterUp) + " | -" + String.valueOf(task.counterDown);
        } else if (newTask.counterUp > 0) {
            streakString = streakString + "+" + String.valueOf(task.counterUp);
        } else if (newTask.counterUp > 0) {
            streakString = streakString + "-" + String.valueOf(task.counterDown);
        }
        if (streakString.length() > 0) {
            streakTextView.setText(streakString);
            streakTextView.setVisibility(View.VISIBLE);
        } else {
            streakTextView.setVisibility(View.GONE);
        }
        super.bindHolder(newTask, position);
    }

    @OnClick(R.id.btnPlus)
    public void onPlusButtonClicked() {
        HabitScoreEvent event = new HabitScoreEvent();
        event.Up = true;
        event.habit = task;
        EventBus.getDefault().post(event);
    }

    @OnClick(R.id.btnMinus)
    public void onMinusButtonClicked() {
        HabitScoreEvent event = new HabitScoreEvent();
        event.Up = false;
        event.habit = task;
        EventBus.getDefault().post(event);
    }

    @Override
    public void setDisabled(boolean openTaskDisabled, boolean taskActionsDisabled) {
        super.setDisabled(openTaskDisabled, taskActionsDisabled);

        this.btnPlus.setEnabled(!taskActionsDisabled);
        this.btnMinus.setEnabled(!taskActionsDisabled);
    }

    @Override
    protected Boolean getTaskIconWrapperIsVisible() {
        Boolean isVisible = super.getTaskIconWrapperIsVisible();
        if (this.streakTextView.getVisibility() == View.VISIBLE) {
            isVisible = true;
        }
        return isVisible;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/viewHolders/tasks/HabitViewHolder.kt
package com.habitrpg.android.habitica.ui.viewHolders.tasks

import android.graphics.PorterDuff
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import androidx.core.content.ContextCompat
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.helpers.GroupPlanInfoProvider
import com.habitrpg.android.habitica.models.tasks.Task
import com.habitrpg.shared.habitica.models.responses.TaskDirection

class HabitViewHolder(
    itemView: View,
    scoreTaskFunc: ((Task, TaskDirection) -> Unit),
    openTaskFunc: ((Pair<Task, View>) -> Unit),
    brokenTaskFunc: ((Task) -> Unit),
    assignedTextProvider: GroupPlanInfoProvider?
) : BaseTaskViewHolder(itemView, scoreTaskFunc, openTaskFunc, brokenTaskFunc, assignedTextProvider) {

    private val btnPlusWrapper: FrameLayout = itemView.findViewById(R.id.btnPlusWrapper)
    private val btnPlusIconView: ImageView = itemView.findViewById(R.id.btnPlusIconView)
    private val btnPlusCircleView: View = itemView.findViewById(R.id.button_plus_circle_view)
    private val btnPlus: Button = itemView.findViewById(R.id.btnPlus)
    private val btnMinusWrapper: FrameLayout = itemView.findViewById(R.id.btnMinusWrapper)
    private val btnMinusIconView: ImageView = itemView.findViewById(R.id.btnMinusIconView)
    private val btnMinusCircleView: View = itemView.findViewById(R.id.button_minus_circle_view)
    private val btnMinus: Button = itemView.findViewById(R.id.btnMinus)

    init {
        btnPlus.setOnClickListener { onPlusButtonClicked() }
        btnPlus.isClickable = true
        btnMinus.setOnClickListener { onMinusButtonClicked() }
        btnMinus.isClickable = true
    }

    override fun bind(
        data: Task,
        position: Int,
        displayMode: String,
        ownerID: String?
    ) {
        this.task = data
        if (data.up == true) {
            val plusIcon = if (isLocked) {
                val icon = ContextCompat.getDrawable(context, R.drawable.task_lock)
                icon?.setTint(ContextCompat.getColor(context, data.darkestTaskColor))
                icon
            } else {
                val icon = ContextCompat.getDrawable(context, R.drawable.habit_plus)
                icon?.setTint(ContextCompat.getColor(context, R.color.white))
                icon
            }
            plusIcon?.setTintMode(PorterDuff.Mode.MULTIPLY)
            this.btnPlusIconView.setImageDrawable(plusIcon)
            val drawable = ContextCompat.getDrawable(context, R.drawable.habit_circle)
            this.btnPlusWrapper.setBackgroundResource(data.lightTaskColor)
            drawable?.setTint(ContextCompat.getColor(context, data.mediumTaskColor))
            drawable?.setTintMode(PorterDuff.Mode.MULTIPLY)
            btnPlusCircleView.background = drawable
            this.btnPlus.visibility = View.VISIBLE
            this.btnPlus.isClickable = true
        } else {
            this.btnPlusWrapper.setBackgroundResource(R.color.habit_inactive_gray)
            val plusIcon = if (isLocked) {
                val icon = ContextCompat.getDrawable(context, R.drawable.task_lock)
                icon?.setTint(ContextCompat.getColor(context, R.color.text_dimmed))
                icon
            } else {
                val icon = ContextCompat.getDrawable(context, R.drawable.habit_plus)
                icon?.setTint(ContextCompat.getColor(context, R.color.content_background_offset))
                icon
            }
            plusIcon?.setTintMode(PorterDuff.Mode.MULTIPLY)
            this.btnPlusIconView.setImageDrawable(plusIcon)
            btnPlusCircleView.background = ContextCompat.getDrawable(context, R.drawable.habit_circle_disabled)
            this.btnPlus.visibility = View.GONE
            this.btnPlus.isClickable = false
        }

        if (data.down == true) {
            this.btnMinusWrapper.setBackgroundResource(data.lightTaskColor)
            val minusIcon = if (isLocked) {
                val icon = ContextCompat.getDrawable(context, R.drawable.task_lock)
                icon?.setTint(ContextCompat.getColor(context, data.darkestTaskColor))
                icon
            } else {
                val icon = ContextCompat.getDrawable(context, R.drawable.habit_minus)
                icon?.setTint(ContextCompat.getColor(context, R.color.white))
                icon
            }
            minusIcon?.setTintMode(PorterDuff.Mode.MULTIPLY)
            this.btnMinusIconView.setImageDrawable(minusIcon)
            val drawable = ContextCompat.getDrawable(context, R.drawable.habit_circle)
            this.btnMinusWrapper.setBackgroundResource(data.lightTaskColor)
            drawable?.setTint(ContextCompat.getColor(context, data.mediumTaskColor))
            drawable?.setTintMode(PorterDuff.Mode.MULTIPLY)
            btnMinusCircleView.background = drawable
            this.btnMinus.visibility = View.VISIBLE
            this.btnMinus.isClickable = true
        } else {
            this.btnMinusWrapper.setBackgroundResource(R.color.habit_inactive_gray)
            val minusIcon = if (isLocked) {
                val icon = ContextCompat.getDrawable(context, R.drawable.task_lock)
                icon?.setTint(ContextCompat.getColor(context, R.color.content_background_offset))
                icon
            } else {
                val icon = ContextCompat.getDrawable(context, R.drawable.habit_minus)
                icon?.setTint(ContextCompat.getColor(context, R.color.content_background_offset))
                icon
            }
            minusIcon?.setTintMode(PorterDuff.Mode.MULTIPLY)
            this.btnMinusIconView.setImageDrawable(minusIcon)
            btnMinusCircleView.background = ContextCompat.getDrawable(context, R.drawable.habit_circle_disabled)
            this.btnMinus.visibility = View.GONE
            this.btnMinus.isClickable = false
        }

        val streakString = task?.streakString
        if (streakString?.isNotEmpty() == true) {
            streakTextView.text = streakString
            streakTextView.visibility = View.VISIBLE
            streakIconView.visibility = View.VISIBLE
        } else {
            streakTextView.visibility = View.GONE
            streakIconView.visibility = View.GONE
        }
        reminderTextView.visibility = View.GONE
        calendarIconView?.visibility = View.GONE
        super.bind(data, position, displayMode, ownerID)
        if (data.up == false && data.down == false) {
            titleTextView.setTextColor(ContextCompat.getColor(context, R.color.text_quad))
            notesTextView?.setTextColor(ContextCompat.getColor(context, R.color.text_quad))
        }
    }

    override fun onLeftActionTouched() {
        super.onLeftActionTouched()
        onPlusButtonClicked()
    }

    override fun onRightActionTouched() {
        super.onRightActionTouched()
        onMinusButtonClicked()
    }

    private fun onPlusButtonClicked() {
        if (task?.up != true) return
        task?.let { scoreTaskFunc.invoke(it, TaskDirection.UP) }
    }

    private fun onMinusButtonClicked() {
        if (task?.down != true) return
        task?.let { scoreTaskFunc.invoke(it, TaskDirection.DOWN) }
    }

    override fun setDisabled(openTaskDisabled: Boolean, taskActionsDisabled: Boolean) {
        super.setDisabled(openTaskDisabled, taskActionsDisabled)

        this.btnPlus.isEnabled = !taskActionsDisabled
        this.btnMinus.isEnabled = !taskActionsDisabled
    }
}
